pkg/validate/security_context_linux.go

```diff
@@ -639,6 +589,151 @@ var _ = framework.KubeDescribe("Security Context", func() {
+				}, time.Minute, time.Second*4).Should(Succeed())
+
+				// $ id -G
+				// 1000 1234 50000
+				expectedOutput := fmt.Sprintf("%d %d %d\n", imagePredefinedGroupUID, supplementalGroup, imagePredefinedGroupGID)
+
+				By("verify groups for the first process of the container")
+				verifyLogContents(podConfig, logPath, expectedOutput, stdoutType)
+
+				By("verify groups for 'exec'-ed process of container")
+				command := []string{"id", "-G"}
+				o := execSyncContainer(rc, containerID, command)
+				Expect(o).To(BeEquivalentTo(expectedOutput))
+			})
+		})
+		When("SupplementalGroupsPolicy=Strict", func() {
+			It("even if the container's primary UID belongs to some groups in the image, runtime should not add SupplementalGroups to them", func() {
+				By("create pod")
+				podID, podConfig, podLogDir = createPodSandboxWithLogDirectory(rc)
+
+				By("create container for security context SupplementalGroups")
+				supplementalGroup := int64(1234)
+				containerName := "container-with-SupplementalGroupsPolicyMerge-" + framework.NewUUID()
+				logPath := containerName + ".log"
+				containerConfig := &runtimeapi.ContainerConfig{
+					Metadata: framework.BuildContainerMetadata(containerName, framework.DefaultAttempt),
+					Image:    &runtimeapi.ImageSpec{Image: testImagePreDefinedGroup},
+					Command:  []string{"sh", "-c", "id -G; while :; do sleep 1; done"},
+					Linux: &runtimeapi.LinuxContainerConfig{
+						SecurityContext: &runtimeapi.LinuxContainerSecurityContext{
+							RunAsUser:                &runtimeapi.Int64Value{Value: imagePredefinedGroupUID},
+							SupplementalGroups:       []int64{supplementalGroup},
+							SupplementalGroupsPolicy: runtimeapi.SupplementalGroupsPolicy_Strict,
+						},
+					},
+					LogPath: logPath,
+				}
+				containerID := framework.CreateContainer(rc, ic, containerConfig, podID, podConfig)
+
+				By("start container")
+				startContainer(rc, containerID)
+
+				Eventually(func(g Gomega) {
+					containerStatus := getContainerStatus(rc, containerID)
+					g.Expect(containerStatus.State).To(Equal(runtimeapi.ContainerState_CONTAINER_RUNNING))
+					// In testImagePreDefinedGroup,
+					// - its default user is default-user(uid=1000)
+					// - default-user belongs to group-defined-in-image(gid=50000) in /etc/group
+					// And, SupplementalGroupsPolicy is Strict
+					//
+					// Thus, firstly attached process identity of the first container processes should be
+					// (5000(defined in /etc/group) is not appended to supplementary groups)
+					// - uid: 1000 (RunAsUser)
+					// - gid: 1000 (default group for uid=1000)
+					// - supplementary groups
+					//   - 1000: self
+					//   - 1234: SupplementalGroups
+					g.Expect(containerStatus.User).To(PointTo(MatchFields(IgnoreExtras, Fields{
+						"Linux": PointTo(MatchFields(IgnoreExtras, Fields{
+							"Uid": Equal(imagePredefinedGroupUID),
+							"Gid": Equal(imagePredefinedGroupUID),
+							// we can not assume the order of gids
```

Comment:
Would sorting the list make it a little more deterministic?

Reply:
Sure, fixed in [c9678f4](https://github.com/kubernetes-sigs/cri-tools/pull/1438/commits/c9678f49ce66db0041b22d4ee59805df679ec169). Sorting removes gstruct usage, too👍.